datafusion/functions-aggregate/src/count.rs

```diff
@@ -752,10 +761,245 @@ impl Accumulator for DistinctCountAccumulator {
     }
 }
 
+/// GroupsAccumulator for COUNT DISTINCT operations
+#[derive(Debug)]
+pub struct DistinctCountGroupsAccumulator {
```

Comment:
As a follow-up, this could be specialized for types as well (e.g. `PrimitveDistinctCountGroupsAccumulator`)

Reply:
Also using the `HashTable` API would probably give some further gains 
https://docs.rs/hashbrown/latest/hashbrown/struct.HashTable.html